Add purchase task to experimental task lists

joshuatf opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds the missing purchase task to the 2 experimental task lists.

Closes #33125  .

### How to test the changes in this Pull Request:

1. Proceed through OBW, and select the paid theme "Blooms" on the final step.
1. Be sure to opt into Marketing^
1. When brought to the Homescreen, notice that the "Add Blooms to my store" task appears.
1. Install WCAdmin Test Helper plugin, and use it to enable `woocommerce_tasklist_setup_experiment_1_*`
1. Refresh the Homescreen.
1. Note that the purchase task exists
2. Enable the `woocommerce_tasklist_setup_experiment_2_*` experiment
3. Refresh the homescreen.
4. Note the purchase task exists

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Add purchase task to experimental task lists

joshuatf committed:
Add changelog entry

louwie17 reviewed (changes requested):
Overall tested well, but left some comments in relation to styling and copy that should match the design of the task list. I think the copy change mentioned at the end here, would be the biggest item, we could do without the header (or do that as part of a separate PR).

For experiment 1 we have headers for each task, but don't have a header for this one, I think we would want to create one.
<img width="593" alt="Screen Shot 2022-05-25 at 4 06 30 PM" src="https://user-images.githubusercontent.com/2240960/170348811-fa60f611-d4f9-4714-8d9c-36f22409fa80.png">
vs
the header of ways to get paid:
<img width="678" alt="Screen Shot 2022-05-25 at 4 07 22 PM" src="https://user-images.githubusercontent.com/2240960/170348863-c594d1cb-ab1f-44fe-8df8-9e4ae7aa088e.png">
@jarekmorawski might have some suggestions?

Also we want to update the titles to match the copy of the new task list.
For example for both experiments, the completed copy says:
'Add paid extensions to my store' but all the others are in past tense, maybe we should change it to 'You added paid extensions to your store'?
See an example of the logic here: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Products.php#L43-L49

jarekmorawski commented:
> 'Add paid extensions to my store' but all the others are in past tense, maybe we should change it to 'You added paid extensions to your store'?

Absolutely! I added the updated designs here: G9vi4e286yjdzyIdg8hBb1-fi-1761%3A159201

Here's how the copy would go:

**Task title**
* `Add %1$s and %2$s` (incomplete)
* `You added %1$s and %2$s` (complete)

**Description**
* `Continue to checkout to to add amazing new features to your store.`

For the experiment 1 (single task list), the button should say `Continue`. Let me know if this helps!

joshuatf committed:
Add purchase header

joshuatf committed:
Update logic around purchase products list strings

joshuatf committed:
Add purchase illustration

joshuatf commented:
Thanks for all the input and help on this, @jarekmorawski!

@louwie17 I've updated based on the discussion in p1653660464272079-slack.  Also updated the width/height of illustrations in https://github.com/woocommerce/woocommerce/pull/33178/commits/4044869c7e766bd1a342c0a6cdb0363b4514f29d which needs a sanity check, but I think will confine proportions well.

louwie17 commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Purchase.php:
```diff
@@ -54,20 +54,52 @@ public function get_id() {
+		$additional_count = count( $products['remaining'] ) - 1;
+
+		if ( $this->get_parent_option( 'use_completed_title' ) && $this->is_complete() ) {
+			return count( $products['remaining'] ) === 1
```
I think this should be `purchaseable` right? As the `remaining` array will always be empty if this task is completed.
As currently I am getting this weird title:
<img width="737" alt="Screen Shot 2022-06-01 at 3 43 14 PM" src="https://user-images.githubusercontent.com/2240960/171482076-b8338cca-9297-4166-9eaf-96fc1e83b734.png">

louwie17 replied:
Same probably goes for $addtional_count and line 79. 
For additional_count probably need a `$completed_additional_count`.

joshuatf replied:
Ya, good point.  This doesn't work well as you complete purchases.  Updated.

louwie17 commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Purchase.php:
```diff
@@ -54,20 +54,52 @@ public function get_id() {
+						$additional_count,
+						'woocommerce'
+					),
+					$products['remaining'][0],
```
Also in the $products['purchaseable'] the items are array's of plugins, so this will probably be `$products['purchaseable'][0]['label']`

louwie17 commented on plugins/woocommerce-admin/client/two-column-tasks/headers/purchase.js:
```diff
@@ -0,0 +1,47 @@
+				<Button
+					isSecondary={ task.isComplete }
+					isPrimary={ ! task.isComplete }
+					onClick={ goToTask }
```
This should actually do something different, as it navigates to the `purchase` task screen which doesn't exist.
It should show a popup like this instead:
<img width="986" alt="Screen Shot 2022-06-01 at 4 07 39 PM" src="https://user-images.githubusercontent.com/2240960/171482961-33d7b2d5-144a-4d84-ac2f-7aed63200451.png">

joshuatf replied:
👍  I think ultimately if we keep this variant of the task list we should strive to dry up these headers and actions a bit.  There is a lot of duplication between components and even backend data being replicated on the frontend that is going to cause headaches down the road.

louwie17 commented on plugins/woocommerce-admin/client/two-column-tasks/headers/purchase.js:
```diff
@@ -0,0 +1,47 @@
+				<Button
+					isSecondary={ task.isComplete }
+					isPrimary={ ! task.isComplete }
+					onClick={ goToTask }
```
See the custom purchase task item we have here: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/tasks/fills/purchase.js

I know for other headers like WooCommerce Payments we copied that custom logic to the header also instead of using `goToTask` -> https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/two-column-tasks/headers/woocommerce-payments.js#L86

louwie17 commented on plugins/woocommerce-admin/client/two-column-tasks/style.scss:
```diff
@@ -241,6 +240,11 @@
+		img {
+			max-height: 150px;
+			width: auto;
+		}
```
Not sure if this works well for all the tasks in the two column layout, I screwed around with that for a while to, maybe we still want to add a max-width?
<img width="611" alt="Screen Shot 2022-06-01 at 4 15 23 PM" src="https://user-images.githubusercontent.com/2240960/171484373-1bb0eb0c-0b5e-4530-807a-854572f214fb.png">

joshuatf replied:
Add max width and max height.  Seems to be working well on my end, can you double check on yours?

louwie17 replied:
Looks good on my end now to :)

louwie17 reviewed (changes requested):
Thanks for the updates Josh, the header looks great, I did notice a minor issue with the CSS changes.
I also left some comments in relation to the action button and the completed copy.

Also sorry if some of these comments show up twice, Github is being annoying....

joshuatf committed:
Fix up task title tests

joshuatf commented:
Hey @louwie17, updated that old test and added another for better coverage around titles.  Could you give this another look?

joshuatf committed:
Handle PR feedback

joshuatf committed:
Fix theme title in purchase task

louwie17 reviewed (approved):
Thanks for the updated changes @joshuatf , nice work on this, this tested well and changes look great 👍 💯
